File: packages/contracts/src/schemas/external-job.ts
import {
  type DeepReadonly,
  generatedSchema,
  validateContractSchema,
} from "../internal.js";
import { canonicalJsonStringify } from "../canonical-json.js";
import type {
  PaidSkillExecutorBindingContract,
  PrincipalReferenceContract,
  Sha256DigestContract,
} from "./paid-invocation.js";
import type { ReferenceContract } from "./spine.js";

export type ExternalJobStageContract =
  | "start"
  | "inspect"
  | "materialize"
  | "finalize";

export type ExternalJobStatusContract =
  | "runnable"
  | "waiting_external"
  | "succeeded"
  | "failed"
  | "superseded"
  | "dead_letter";

export type ExternalJobFailureContract = DeepReadonly<{
  code: string;
  message: string;
  retryable: boolean;
}>;

export type ExternalJobCheckpointContract = DeepReadonly<Record<string, unknown>>;

export type ExternalJobContinuationContract = DeepReadonly<{
  continuation_id: string;
  principal_ref: PrincipalReferenceContract;
  vendor_ref: PrincipalReferenceContract;
  invocation_ref: ReferenceContract;
  source_run_ref: ReferenceContract;
  execution_binding: PaidSkillExecutorBindingContract;
  operation_identity: Sha256DigestContract;
  stage: ExternalJobStageContract;
  status: ExternalJobStatusContract;
  attempts: number;
  max_attempts: number;
  next_attempt_at?: string;
  deadline_at: string;
  provider_job_ref?: ReferenceContract;
  result_artifact_ref?: ReferenceContract;
  terminal_evidence_ref?: ReferenceContract;
  terminal_evidence_digest?: Sha256DigestContract;
  failure?: ExternalJobFailureContract;
  created_at: string;
  updated_at: string;
}>;

export type ExternalJobScheduleContract = DeepReadonly<{
  continuation_id: string;
  principal_ref: PrincipalReferenceContract;
  vendor_ref: PrincipalReferenceContract;
  invocation_ref: ReferenceContract;
  source_run_ref: ReferenceContract;
  execution_binding: PaidSkillExecutorBindingContract;
  operation_identity: Sha256DigestContract;
  checkpoint: ExternalJobCheckpointContract;
  max_attempts: number;
  next_attempt_at: string;
  deadline_at: string;
  created_at: string;
}>;

export type ExternalJobScheduleIntentContract = DeepReadonly<{
  schema: "runx.external_job_schedule_intent.v1";
  checkpoint: ExternalJobCheckpointContract;
  max_attempts: number;
  initial_delay_ms: number;
  deadline_ms: number;
}>;

export type ExternalJobStageRequestContract = DeepReadonly<{
  continuation: ExternalJobContinuationContract;
  checkpoint: ExternalJobCheckpointContract;
  operation_key: Sha256DigestContract;
}>;

export type ExternalJobStageResultContract = DeepReadonly<
  | {
      status: "waiting";
      provider_job_ref: ReferenceContract;
      checkpoint: ExternalJobCheckpointContract;
      retry_after_ms: number;
    }
  | {
      status: "materialize";
      provider_job_ref: ReferenceContract;
      checkpoint: ExternalJobCheckpointContract;
      retry_after_ms: number;
    }
  | {
      status: "ready";
      provider_job_ref: ReferenceContract;
      result_artifact_ref: ReferenceContract;
      evidence_ref: ReferenceContract;
      evidence_digest: Sha256DigestContract;
    }
  | {
      status: "provider_failed";
      provider_job_ref: ReferenceContract;
      evidence_ref: ReferenceContract;
      evidence_digest: Sha256DigestContract;
      failure: ExternalJobFailureContract;
    }
>;

export const externalJobContinuationV1Schema =
  generatedSchema<ExternalJobContinuationContract>("external-job-continuation.schema.json");
export const externalJobScheduleV1Schema =
  generatedSchema<ExternalJobScheduleContract>("external-job-schedule.schema.json");
export const externalJobScheduleIntentV1Schema =
  generatedSchema<ExternalJobScheduleIntentContract>("external-job-schedule-intent.schema.json");
export const externalJobStageRequestV1Schema =
  generatedSchema<ExternalJobStageRequestContract>("external-job-stage-request.schema.json");
export const externalJobStageResultV1Schema =
  generatedSchema<ExternalJobStageResultContract>("external-job-stage-result.schema.json");

export function validateExternalJobContinuationContract(
  value: unknown,
  label = "external_job_continuation",
): ExternalJobContinuationContract {
  const continuation = validateContractSchema(externalJobContinuationV1Schema, value, label);
  validateSemantics(continuation, label);
  return continuation;
}

export function validateExternalJobScheduleContract(
  value: unknown,
  label = "external_job_schedule",
): ExternalJobScheduleContract {
  const schedule = validateContractSchema(externalJobScheduleV1Schema, value, label);
  validateIdentityReferences(schedule, label);
  checkpoint(schedule.checkpoint, `${label}.checkpoint`);
  const createdAt = timestamp(schedule.created_at, `${label}.created_at`);
  const nextAttemptAt = timestamp(schedule.next_attempt_at, `${label}.next_attempt_at`);
  const deadlineAt = timestamp(schedule.deadline_at, `${label}.deadline_at`);
  if (nextAttemptAt < createdAt || nextAttemptAt > deadlineAt || deadlineAt <= createdAt) {
    throw new Error(`${label} timestamps are not ordered.`);
  }
  return schedule;
}

export function validateExternalJobScheduleIntentContract(
  value: unknown,
  label = "external_job_schedule_intent",
): ExternalJobScheduleIntentContract {
  const intent = validateContractSchema(externalJobScheduleIntentV1Schema, value, label);
  checkpoint(intent.checkpoint, `${label}.checkpoint`);
  return intent;
}

export function validateExternalJobStageRequestContract(
  value: unknown,
  label = "external_job_stage_request",
): ExternalJobStageRequestContract {
  const request = validateContractSchema(externalJobStageRequestV1Schema, value, label);
  validateExternalJobContinuationContract(request.continuation, `${label}.continuation`);
  checkpoint(request.checkpoint, `${label}.checkpoint`);
  return request;
}

export function validateExternalJobStageResultContract(
  value: unknown,
  label = "external_job_stage_result",
): ExternalJobStageResultContract {
  const result = validateContractSchema(externalJobStageResultV1Schema, value, label);
  referenceType(result.provider_job_ref, "target", `${label}.provider_job_ref`);
  if (result.status === "waiting" || result.status === "materialize") {
    checkpoint(result.checkpoint, `${label}.checkpoint`);
  } else {
    referenceType(result.evidence_ref, "verification", `${label}.evidence_ref`);
    if (result.status === "ready") {
      referenceType(result.result_artifact_ref, "artifact", `${label}.result_artifact_ref`);
    } else if (result.failure.retryable) {
      throw new Error(`${label}.provider_failed must be terminal.`);
    }
  }
  return result;
}

function validateSemantics(
  continuation: ExternalJobContinuationContract,
  label: string,
): void {
  validateIdentityReferences(continuation, label);
  if (
    !Number.isSafeInteger(continuation.attempts)
    || continuation.attempts < 0
    || continuation.attempts > continuation.max_attempts
  ) {
    throw new Error(`${label}.attempts must be a safe integer between zero and max_attempts.`);
  }
  const createdAt = timestamp(continuation.created_at, `${label}.created_at`);
  const updatedAt = timestamp(continuation.updated_at, `${label}.updated_at`);
  const deadlineAt = timestamp(continuation.deadline_at, `${label}.deadline_at`);
  if (updatedAt < createdAt || deadlineAt <= createdAt) {
    throw new Error(`${label} timestamps are not ordered.`);
  }
  if (continuation.next_attempt_at) {
    const nextAttemptAt = timestamp(continuation.next_attempt_at, `${label}.next_attempt_at`);
    if (nextAttemptAt < createdAt || nextAttemptAt > deadlineAt) {
      throw new Error(`${label}.next_attempt_at must fall within the continuation deadline.`);
    }
  }

  const terminal = ["succeeded", "failed", "superseded", "dead_letter"]
    .includes(continuation.status);
  if (terminal === Boolean(continuation.next_attempt_at)) {
    throw new Error(`${label} must schedule exactly non-terminal state.`);
  }
  if (continuation.status === "waiting_external") {
    if (continuation.stage !== "inspect" || !continuation.provider_job_ref) {
      throw new Error(`${label} waiting_external state requires inspect and a provider job reference.`);
    }
  } else if (continuation.status === "runnable") {
    if (continuation.stage === "inspect") {
      throw new Error(`${label} inspect state must be waiting_external.`);
    }
    if (continuation.stage !== "start" && !continuation.provider_job_ref) {
      throw new Error(`${label} post-start state requires a provider job reference.`);
    }
  }

  if (!terminal) {
    if (
      continuation.result_artifact_ref
      || continuation.terminal_evidence_ref
      || continuation.terminal_evidence_digest
      || continuation.failure
    ) {
      throw new Error(`${label} active state cannot expose terminal fields.`);
    }
    return;
  }
  if (continuation.status === "succeeded") {
    if (
      continuation.stage !== "finalize"
      || !continuation.result_artifact_ref
      || !continuation.terminal_evidence_ref
      || !continuation.terminal_evidence_digest
      || continuation.failure
    ) {
      throw new Error(`${label} succeeded state requires finalized artifact and evidence.`);
    }
    return;
  }
  if (continuation.result_artifact_ref) {
    throw new Error(`${label} non-success terminal state cannot expose a result artifact.`);
  }
  if (continuation.status === "dead_letter") {
    if (!continuation.failure || continuation.failure.retryable) {
      throw new Error(`${label} dead_letter state requires a non-retryable bounded failure.`);
    }
    return;
  }
  if (!continuation.terminal_evidence_ref || !continuation.terminal_evidence_digest) {
    throw new Error(`${label} terminal provider outcome requires exact evidence.`);
  }
  if (continuation.status === "failed" && (!continuation.failure || continuation.failure.retryable)) {
    throw new Error(`${label} failed state requires a terminal failure.`);
  }
  if (continuation.status === "superseded" && continuation.failure) {
    throw new Error(`${label} superseded state cannot claim provider failure.`);
  }
}

function validateIdentityReferences(
  value: Pick<
    ExternalJobContinuationContract,
    "invocation_ref" | "source_run_ref"
  >,
  label: string,
): void {
  referenceType(value.invocation_ref, "target", `${label}.invocation_ref`);
  if (!value.invocation_ref.uri.startsWith("runx:paid-invocation:")) {
    throw new Error(`${label}.invocation_ref must identify a paid invocation.`);
  }
  referenceType(value.source_run_ref, "act", `${label}.source_run_ref`);
  if (!value.source_run_ref.uri.startsWith("runx:run:")) {
    throw new Error(`${label}.source_run_ref must identify a hosted run.`);
  }
}

function referenceType(
  reference: ReferenceContract,
  type: ReferenceContract["type"],
  label: string,
): void {
  if (reference.type !== type) throw new Error(`${label}.type must be ${type}.`);
}

function timestamp(value: string, label: string): number {
  const parsed = Date.parse(value);
  if (!Number.isFinite(parsed)) throw new Error(`${label} must be a valid timestamp.`);
  return parsed;
}

function checkpoint(value: ExternalJobCheckpointContract, label: string): void {
  const bytes = new TextEncoder().encode(canonicalJsonStringify(value)).byteLength;
  if (bytes > 64 * 1024) throw new Error(`${label} exceeds 65536 bytes.`);
}

File: crates/runx-contracts/src/external_job.rs
//! Portable durable-external-job continuation contract.
//!
//! The contract records provider-neutral execution progress only. Hosted
//! persistence, leases, credentials, SDK values, and provider response bodies
//! stay behind Cloud adapters.

use serde::de::{self, Deserializer};
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};
use std::collections::BTreeMap;

use crate::schema::{BoundedString, IsoDateTime, RunxSchema};
use crate::{PaidSkillExecutorBinding, PrincipalReference, Reference, Sha256Digest};

pub const EXTERNAL_JOB_CONTINUATION_SCHEMA: &str = "runx.external_job_continuation.v1";
pub const EXTERNAL_JOB_SCHEDULE_SCHEMA: &str = "runx.external_job_schedule.v1";
pub const EXTERNAL_JOB_SCHEDULE_INTENT_SCHEMA: &str = "runx.external_job_schedule_intent.v1";
pub const EXTERNAL_JOB_STAGE_REQUEST_SCHEMA: &str = "runx.external_job_stage.request.v1";
pub const EXTERNAL_JOB_STAGE_RESULT_SCHEMA: &str = "runx.external_job_stage.result.v1";
const EXTERNAL_JOB_CHECKPOINT_MAX_BYTES: usize = 64 * 1024;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(rename_all = "snake_case")]
pub enum ExternalJobStage {
    Start,
    Inspect,
    Materialize,
    Finalize,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(rename_all = "snake_case")]
pub enum ExternalJobStatus {
    Runnable,
    WaitingExternal,
    Succeeded,
    Failed,
    Superseded,
    DeadLetter,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum ExternalJobScheduleIntentSchema {
    #[serde(rename = "runx.external_job_schedule_intent.v1")]
    V1,
}

/// Bounded for durable scheduling while allowing multi-hour provider jobs to
/// use a five-minute steady-state inspection cadence.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(transparent)]
pub struct ExternalJobAttemptLimit(u8);

impl ExternalJobAttemptLimit {
    pub fn new(value: u8) -> Option<Self> {
        (1..=128).contains(&value).then_some(Self(value))
    }

    pub fn get(self) -> u8 {
        self.0
    }
}

impl<'de> Deserialize<'de> for ExternalJobAttemptLimit {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let value = u8::deserialize(deserializer)?;
        Self::new(value).ok_or_else(|| de::Error::custom("max_attempts must be between 1 and 128"))
    }
}

impl RunxSchema for ExternalJobAttemptLimit {
    fn json_schema() -> Value {
        json!({ "type": "integer", "minimum": 1, "maximum": 128 })
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(transparent)]
pub struct ExternalJobDelayMillis(u32);

impl<'de> Deserialize<'de> for ExternalJobDelayMillis {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let value = u32::deserialize(deserializer)?;
        (value <= 300_000).then_some(Self(value)).ok_or_else(|| {
            de::Error::custom("external job delay must not exceed 300000 milliseconds")
        })
    }
}

impl RunxSchema for ExternalJobDelayMillis {
    fn json_schema() -> Value {
        json!({ "type": "integer", "minimum": 0, "maximum": 300_000 })
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(transparent)]
pub struct ExternalJobDeadlineMillis(u32);

impl<'de> Deserialize<'de> for ExternalJobDeadlineMillis {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let value = u32::deserialize(deserializer)?;
        (1_000..=86_400_000)
            .contains(&value)
            .then_some(Self(value))
            .ok_or_else(|| {
                de::Error::custom(
                    "external job deadline must be between 1000 and 86400000 milliseconds",
                )
            })
    }
}

impl RunxSchema for ExternalJobDeadlineMillis {
    fn json_schema() -> Value {
        json!({ "type": "integer", "minimum": 1_000, "maximum": 86_400_000 })
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct ExternalJobFailure {
    pub code: BoundedString<96>,
    pub message: BoundedString<500>,
    pub retryable: bool,
}

/// Private package state only. Credentials, SDK values, presigned URLs, and
/// raw provider responses never belong here.
#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
#[serde(transparent)]
pub struct ExternalJobCheckpoint(BTreeMap<String, Value>);

impl<'de> Deserialize<'de> for ExternalJobCheckpoint {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let value = BTreeMap::<String, Value>::deserialize(deserializer)?;
        let encoded = serde_json::to_vec(&value).map_err(de::Error::custom)?;
        if encoded.len() > EXTERNAL_JOB_CHECKPOINT_MAX_BYTES {
            return Err(de::Error::custom(
                "external job checkpoint exceeds 65536 bytes",
            ));
        }
        Ok(Self(value))
    }
}

impl RunxSchema for ExternalJobCheckpoint {
    fn json_schema() -> Value {
        // Package checkpoints are intentionally open, but the openness must be
        // explicit so packet-schema drift checks cannot mistake it for an
        // unspecified envelope shape.
        json!({ "type": "object", "additionalProperties": true })
    }
}

/// Current V1 continuation state. `schema` is injected by the generated JSON
/// Schema identity and is optional on the Rust value, matching other Runx
/// packet contracts.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.external_job_continuation.v1")]
pub struct ExternalJobContinuation {
    pub continuation_id: BoundedString<256>,
    pub principal_ref: PrincipalReference,
    pub vendor_ref: PrincipalReference,
    pub invocation_ref: Reference,
    pub source_run_ref: Reference,
    pub execution_binding: PaidSkillExecutorBinding,
    pub operation_identity: Sha256Digest,
    pub stage: ExternalJobStage,
    pub status: ExternalJobStatus,
    pub attempts: u32,
    pub max_attempts: ExternalJobAttemptLimit,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_attempt_at: Option<IsoDateTime>,
    pub deadline_at: IsoDateTime,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider_job_ref: Option<Reference>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result_artifact_ref: Option<Reference>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub terminal_execution_receipt_ref: Option<Reference>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub terminal_evidence_ref: Option<Reference>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub terminal_evidence_digest: Option<Sha256Digest>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub failure: Option<ExternalJobFailure>,
    pub created_at: IsoDateTime,
    pub updated_at: IsoDateTime,
}

/// Durable outbox payload used to create one continuation exactly once.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.external_job_schedule.v1")]
pub struct ExternalJobSchedule {
    pub continuation_id: BoundedString<256>,
    pub principal_ref: PrincipalReference,
    pub vendor_ref: PrincipalReference,
    pub invocation_ref: Reference,
    pub source_run_ref: Reference,
    pub execution_binding: PaidSkillExecutorBinding,
    pub operation_identity: Sha256Digest,
    pub checkpoint: ExternalJobCheckpoint,
    pub max_attempts: ExternalJobAttemptLimit,
    pub next_attempt_at: IsoDateTime,
    pub deadline_at: IsoDateTime,
    pub created_at: IsoDateTime,
}

/// Product-owned opt-in emitted by the initial execution package. Trusted
/// invocation identity and absolute scheduling timestamps are added by Runx.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.external_job_schedule_intent.v1")]
pub struct ExternalJobScheduleIntent {
    pub schema: ExternalJobScheduleIntentSchema,
    /// Runner in the same immutable package that owns continuation stages.
    /// Runx resolves and pins its full execution closure before scheduling.
    pub stage_runner: BoundedString<128>,
    pub checkpoint: ExternalJobCheckpoint,
    pub max_attempts: ExternalJobAttemptLimit,
    pub initial_delay_ms: ExternalJobDelayMillis,
    pub deadline_ms: ExternalJobDeadlineMillis,
}

/// Exact package re-entry input for one persisted stage.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.external_job_stage.request.v1")]
pub struct ExternalJobStageRequest {
    pub continuation: ExternalJobContinuation,
    pub checkpoint: ExternalJobCheckpoint,
    pub operation_key: Sha256Digest,
}

/// Provider-neutral package result. Runx owns scheduling and finalization;
/// the package owns only the meaning of each stage.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(tag = "status", rename_all = "snake_case", deny_unknown_fields)]
#[runx_schema(id = "runx.external_job_stage.result.v1")]
pub enum ExternalJobStageResult {
    Waiting {
        provider_job_ref: Reference,
        checkpoint: ExternalJobCheckpoint,
        retry_after_ms: ExternalJobDelayMillis,
    },
    Materialize {
        provider_job_ref: Reference,
        checkpoint: ExternalJobCheckpoint,
        retry_after_ms: ExternalJobDelayMillis,
    },
    Ready {
        provider_job_ref: Reference,
        result_artifact_ref: Reference,
        evidence_ref: Reference,
        evidence_digest: Sha256Digest,
    },
    ProviderFailed {
        provider_job_ref: Reference,
        evidence_ref: Reference,
        evidence_digest: Sha256Digest,
        failure: ExternalJobFailure,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn attempt_limit_is_bounded() {
        assert!(ExternalJobAttemptLimit::new(1).is_some());
        assert!(ExternalJobAttemptLimit::new(128).is_some());
        assert!(ExternalJobAttemptLimit::new(0).is_none());
        assert!(ExternalJobAttemptLimit::new(129).is_none());
    }
}
